src/local.rs

```diff
@@ -453,6 +455,82 @@ impl ObjectStore for LocalFileSystem {
     }
 }
 
+enum LocalUploadOperation {
+    Write,
+}
+
+enum LocalUploadState {
+    Idle,
+    Busy(JoinHandle<Result<LocalUploadOperation, io::Error>>),
+}
+
+struct LocalUpload {
+    state: LocalUploadState,
+    file: Arc<std::fs::File>,
+}
+
+#[async_trait]
+impl MultiPartUpload for LocalUpload {
+    async fn abort(&mut self) -> Result<()> {
+        Ok(())
+    }
+}
+
+impl AsyncWrite for LocalUpload {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> std::task::Poll<Result<usize, io::Error>> {
+        loop {
+            match &mut self.state {
```

Comment:
I think this could be simplified with maybe_spawn_blocking and `BoxFuture` instead of `JoinHandle`

Reply:
Main reason I didn't use `maybe_spawn_blocking` was it forces the wrong error type for this function, and don't see any easy way to make that generic. But I agree `BoxFuture` would be better.